In [1]:
%cd /nfsd/sysbiobig/bazzaccoen/tangramlit_dev

/nfsd/sysbiobig/bazzaccoen/tangramlit_dev


In [2]:
import tangramlit as tgl

/nfsd/sysbiobig/bazzaccoen/tangramlit_dev/myvenv/lib64/python3.12/site-packages/dask/dataframe/__init__.py:31: FutureWarning: The legacy Dask DataFrame implementation is deprecated and will be removed in a future version. Set the configuration option `dataframe.query-planning` to `True` or None to enable the new Dask Dataframe implementation and silence this warning.
  warnings.warn(
/nfsd/sysbiobig/bazzaccoen/tangramlit_dev/myvenv/lib64/python3.12/site-packages/xarray_schema/__init__.py:1: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  from pkg_resources import DistributionNotFound, get_distribution
/nfsd/sysbiobig/bazzaccoen/tangramlit_dev/myvenv/lib64/python3.12/site-packages/anndata/utils.py:434: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text 

In [3]:
import scanpy as sc
import numpy as np 
import pandas as pd
import anndata as ad
import yaml

# Data preparation

In [4]:
# Set data path
data_path = "/nfsd/sysbiobig/bazzaccoen/tangramlit_dev/data/Dataset3/"

## SC data

In [ ]:
adata_sc = sc.read_10x_mtx(
    data_path + 'GSM4800800_HIPP_sc_Rep1_10X',
    var_names='gene_symbols',
    cache=True
)
adata_sc

AnnData object with n_obs × n_vars = 8653 × 31053
    var: 'gene_ids', 'feature_types'

In [13]:
adata_sc.obs  # no columns

Empty DataFrame
Columns: []
Index: [AAACCTGAGATGCCAG-1, AAACCTGAGGCGCTCT-1, AAACCTGCAACACCTA-1, AAACCTGCATGCCTAA-1, AAACCTGGTCGAACAG-1, AAACCTGGTGGTACAG-1, AAACCTGGTTAAGACA-1, AAACCTGGTTATTCTC-1, AAACCTGGTTTGTTGG-1, AAACCTGTCCTTTACA-1, AAACCTGTCGGCATCG-1, AAACCTGTCTGGCGTG-1, AAACGGGAGCGATCCC-1, AAACGGGAGCTGCCCA-1, AAACGGGAGCTGTCTA-1, AAACGGGAGGTTCCTA-1, AAACGGGAGTACGCCC-1, AAACGGGAGTATCTCG-1, AAACGGGAGTGGAGAA-1, AAACGGGCAATGAATG-1, AAACGGGCACCACGTG-1, AAACGGGCAGTTCATG-1, AAACGGGCATGGTAGG-1, AAACGGGGTACAGTTC-1, AAACGGGGTCACCCAG-1, AAACGGGGTGCATCTA-1, AAACGGGTCCACGACG-1, AAACGGGTCCCAAGTA-1, AAACGGGTCCCTGACT-1, AAACGGGTCGATGAGG-1, AAACGGGTCTCGATGA-1, AAACGGGTCTGCTGTC-1, AAAGATGAGACAGACC-1, AAAGATGAGACTAGAT-1, AAAGATGAGATCTGCT-1, AAAGATGAGATGGCGT-1, AAAGATGAGCCACGTC-1, AAAGATGAGCCGCCTA-1, AAAGATGAGCGTAATA-1, AAAGATGAGCTCCTCT-1, AAAGATGAGGCAAAGA-1, AAAGATGAGTATCGAA-1, AAAGATGAGTGTTAGA-1, AAAGATGCAATGGAAT-1, AAAGATGCACACTGCG-1, AAAGATGCACTAGTAC-1, AAAGATGCATCGGGTC-1, AAAGATGGTAGGACAC-1, AAAGATGGTCTGCCAG-1, AAAGATGGTGCACTTA-1, AAAGATGGTTCAGCGC-1, AAAGATGGTTGGTAAA-1, AAAGATGTCACCGGGT-1, AAAGATGTCATCGATG-1, AAAGATGTCGAATGCT-1, AAAGATGTCGCCTGAG-1, AAAGATGTCTCTGTCG-1, AAAGCAAAGATGGGTC-1, AAAGCAAAGCACACAG-1, AAAGCAAAGGGATCTG-1, AAAGCAAAGGTGATAT-1, AAAGCAAAGGTGTGGT-1, AAAGCAAAGTTTGCGT-1, AAAGCAACAAAGAATC-1, AAAGCAACACATAACC-1, AAAGCAACACTGCCAG-1, AAAGCAACAGTATCTG-1, AAAGCAACATGAGCGA-1, AAAGCAACATGCCTTC-1, AAAGCAACATGCTAGT-1, AAAGCAACATTGGGCC-1, AAAGCAAGTATCGCAT-1, AAAGCAAGTCCAACTA-1, AAAGCAAGTCTCAACA-1, AAAGCAAGTGTAATGA-1, AAAGCAAGTTCAGGCC-1, AAAGCAAGTTCGCGAC-1, AAAGCAATCACGGTTA-1, AAAGCAATCACTCTTA-1, AAAGCAATCGCCGTGA-1, AAAGCAATCTCATTCA-1, AAAGCAATCTGTCTCG-1, AAAGTAGAGCCACGCT-1, AAAGTAGAGCTGCAAG-1, AAAGTAGAGGATGCGT-1, AAAGTAGAGGTTACCT-1, AAAGTAGCAACACCCG-1, AAAGTAGCAAGCCATT-1, AAAGTAGCACCCATTC-1, AAAGTAGCACGTAAGG-1, AAAGTAGCAGCCAATT-1, AAAGTAGCATCTATGG-1, AAAGTAGGTCGCTTCT-1, AAAGTAGGTGATAAAC-1, AAAGTAGTCGTGGTCG-1, AAAGTAGTCTCCGGTT-1, AAATGCCAGACCTTTG-1, AAATGCCAGAGTGACC-1, AAATGCCAGCCAACAG-1, AAATGCCAGCCTATGT-1, ...]

[8653 rows x 0 columns]

In [9]:
adata_sc.var

gene_ids    feature_types
Xkr4            ENSMUSG00000051951  Gene Expression
Gm1992          ENSMUSG00000089699  Gene Expression
Gm37381         ENSMUSG00000102343  Gene Expression
Rp1             ENSMUSG00000025900  Gene Expression
Sox17           ENSMUSG00000025902  Gene Expression
...                            ...              ...
AC168977.1      ENSMUSG00000079808  Gene Expression
AC149090.1      ENSMUSG00000095041  Gene Expression
CAAA01118383.1  ENSMUSG00000063897  Gene Expression
Vmn2r122        ENSMUSG00000096730  Gene Expression
CAAA01147332.1  ENSMUSG00000095742  Gene Expression

[31053 rows x 2 columns]

In [14]:
adata_sc.X

<Compressed Sparse Column sparse matrix of dtype 'float32'
	with 14037406 stored elements and shape (8653, 31053)>

SC data contains only cell IDs (empty obs dataframe), gene symbols/IDs and counts for $8653$ cells $\times 31053$ genes. The benchmarking dataset has $8596 \times 16384$ cells/genes respectively.

It needs dowmnsampling and preprocessing.

In [ ]:
# save h5ad
sc.write(data_path + "scRNA_data.h5ad", adata_sc, 'h5ad')

/tmp/ipykernel_59653/1761400515.py:1: FutureWarning: The specified parameters ('ext',) are no longer positional. Please specify them like `ext='h5ad'`
  sc.write(data_path + "/scRNA_data.h5ad", adata_sc, 'h5ad')


## seqFISH data

In [ ]:
# read xlsx sheets
seqfish_counts = pd.read_excel(data_path + 'spatial_counts_centroids.xlsx', sheet_name='Hippocampus Counts', header=None, index_col=0)
seqfish_counts.index = seqfish_counts.index.str.strip("'\"")  # remove double quoting "''"
seqfish_coords = pd.read_excel(data_path + 'spatial_counts_centroids.xlsx', sheet_name='Centroids', header=None, index_col=None)
seqfish_coords.index = seqfish_coords.index + 1  # align to seqfish_counts.columns
seqfish_coords.columns = ["X", "Y", "Label"]  # name columns


In [33]:
seqfish_counts.head()

1     2     3     4     5     6     7     8     9     10    ...  3576  \
0                                                                   ...         
Tal1      34    18    15    21    17    12    11     4    14    24  ...    13   
Dmbx1     14    15    11    20     1    14     8     9     3     7  ...     2   
Emx2      29    23    32    29     8    13     6     6     9     9  ...     1   
Uncx      17     6     3     6    14     5    11     1     0     7  ...     0   
Paxip1    31    36    42    44    15     8    20    19    13    11  ...     7   

        3577  3578  3579  3580  3581  3582  3583  3584  3585  
0                                                             
Tal1       0     1     7     6     3     0     4     1     4  
Dmbx1      2     2     8     5     1     0     3     2     0  
Emx2       5     4     3     6     0     4     0     6     0  
Uncx       0     3     8     0     0     2     0     4     6  
Paxip1    10     4    19     5     1     1     9     1     3  

[5 rows x 3585 columns]

In [32]:
seqfish_coords.head()


X    Y  Label
1  490  760      2
2   98  532      3
3  872  570      3
4  936  590      3
5  432  722      3

Spatial data contains counts in a $_{genes} \times n_{spots}$ dataframe and spatial ceontroids coordinates in a separate dataframe.

Shape matches the one in the benchmarking paper exactly.

In [35]:
# Create anndata
adata_st = ad.AnnData(
    seqfish_counts.T,  # transpose counts df
    obs=seqfish_coords,
)
adata_st.obsm['spatial'] = adata_st.obs[['X', 'Y']].to_numpy()
adata_st

/nfsd/sysbiobig/bazzaccoen/tangramlit_dev/myvenv/lib64/python3.12/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.
  warnings.warn("Transforming to str index.", ImplicitModificationWarning)


AnnData object with n_obs × n_vars = 3585 × 249
    obs: 'X', 'Y', 'Label'
    obsm: 'spatial'

In [ ]:
# save anndata
sc.write(data_path + "seqFISH_data.h5ad", adata_st, 'h5ad')

Index(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10',
       ...
       '3576', '3577', '3578', '3579', '3580', '3581', '3582', '3583', '3584',
       '3585'],
      dtype='object', length=3585)

# Data Loading

In [11]:
adata_sc = sc.read(data_path + "scRNA_data.h5ad")
adata_sc

AnnData object with n_obs × n_vars = 8653 × 31053
    var: 'gene_ids', 'feature_types'

In [12]:
adata_st = sc.read(data_path + "seqFISH_data.h5ad")
adata_st

AnnData object with n_obs × n_vars = 3585 × 249
    obs: 'X', 'Y', 'Label'
    obsm: 'spatial'

# Config

In [13]:
# Read config yaml
with open("data/Dataset3/train_config.yaml", "r") as f:
        config = yaml.safe_load(f)

config  # no cluster_label and ct_islands

{'cluster_label': None,
 'lambda_count': 1e-05,
 'lambda_ct_islands': 0,
 'lambda_d': 0.001,
 'lambda_f_reg': 1e-05,
 'lambda_g1': 1,
 'lambda_g2': 1,
 'lambda_geary': 1,
 'lambda_getis_ord': 1,
 'lambda_l1': 1e-15,
 'lambda_l2': 1e-18,
 'lambda_moran': 1,
 'lambda_neighborhood_g1': 1,
 'lambda_r': 1e-09,
 'lambda_sparsity_g1': 1,
 'learning_rate': 0.1,
 'filter': False,
 'num_epochs': 1000,
 'random_state': 42,
 'target_count': None}

# Train/val split

In [14]:
# Get shared genes (case-insensitive)
sc_genes = {gene.lower(): gene for gene in adata_sc.var_names}
st_genes = {gene.lower(): gene for gene in adata_st.var_names}

# Find intersection of lowercase gene names
shared_genes_set = set(sc_genes.keys()) & set(st_genes.keys())
shared_genes = [gene_lower for gene_lower in shared_genes_set]

# Shuffle the shared genes
shared_genes = np.array(shared_genes)
np.random.seed(config['random_state'])
np.random.shuffle(shared_genes)

# Split into train and validation
train_ratio = 0.8
n_train = int(len(shared_genes) * train_ratio)
train_genes = shared_genes[:n_train]
val_genes = shared_genes[n_train:]

print(len(train_genes), "training genes: ", train_genes[0:10], "...")
print(len(val_genes), "validation genes: ", val_genes[0:10], "...")

195 training genes:  ['irx4' 'tfdp2' 'myb' 'nr2f2' 'sin3a' 'ttf1' 'barhl1' 'foxc1' 'hoxa1'
 'reln'] ...
49 validation genes:  ['klf1' 'med14' 'elf1' 'foxn4' 'pin1' 'cdc6' 'xdh' 'cdc5l' 'smad3' 'irx5'] ...


# Model training

In [15]:
ad_map, mapper, mapper_data = tgl.map_cells_to_space(
        adata_sc=adata_sc, 
        adata_st=adata_st, 
        train_genes_names=train_genes,
        val_genes_names=val_genes,
        **config,
        )

GPU available: False, used: False
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs



  | Name               | Type      | Params | Mode 
---------------------------------------------------------
0 | _density_criterion | KLDivLoss | 0      | train
  | other params       | n/a       | 31.0 M | n/a  
---------------------------------------------------------
31.0 M    Trainable params
0         Non-trainable params
31.0 M    Total params
124.084   Total estimated model params size (MB)
1         Modules in train mode
0         Modules in eval mode



Validating with 45 genes
S matrix shape: torch.Size([8653, 45])
G matrix shape: torch.Size([3585, 45])

Validation 0: {'val_score': 0.7532742023468018, 'val_sparsity-weighted_score': 0.016755450516939163, 'val_AUC': 0.040464311838150024, 'val_entropy': 0.9390270113945007}


Trainig:   0%|          | 0/1000 [00:00<?, ?it/s]


Training with 185 genes
S matrix shape: torch.Size([8653, 185])
G matrix shape: torch.Size([3585, 185])


: 

In [6]:
# read training results
ad_map = sc.read(filename="/nfsd/sysbiobig/bazzaccoen/tangramlit_dev/results/adata_map_Dataset3.h5ad")

In [10]:
# Plot main loss
tgl.plot_loss_terms(adata_map=ad_map, loss_key=["main_loss", "vg_reg", "kl_reg"])

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

In [ ]:
# Score terms
tgl.plot_loss_terms(adata_map=ad_map, loss_key=["main_loss", "sparsity_term", "neighborhood_term"], lambda_scale=False,
                   make_subplot=True, subplot_shape=(1,3))

In [ ]:
# LISA terms
tgl.plot_loss_terms(adata_map=ad_map, loss_key=["main_loss", "getis_ord_term", "moran_term", "geary_term"], 
                    lambda_scale=False, make_subplot=True)

In [ ]:
# CT islands term
tgl.plot_loss_terms(adata_map=ad_map, loss_key="ct_island_term", lambda_scale=False)

# Validate

In [ ]:
# call trainer.validate()
full_val = tgl.validate_mapping_experiment(mapper, mapper_data)

# Sparsity

In [ ]:
# Project all sc data onto spots
ad_ge = tgl.project_sc_genes_onto_space(ad_map, mapper_data)
ad_ge

In [ ]:
# Create training genes scores dataframe
df = tgl.compare_spatial_gene_expr(ad_ge, mapper_data)

In [ ]:
# Plot training scores panels
tgl.plot_training_scores(df)

In [ ]:
# Plot polyfit on test genes
tgl.plot_auc_curve(df)  # same as validation genes

# Save

In [ ]:
# Write tgl.map_cells_to_space() output to .h5ad
sc.write(filename='/nfsd/sysbiobig/bazzaccoen/tangramlit_dev/results/adata_map_Dataset3', adata=ad_map, ext='h5ad')